### PR TITLE
Add support for springs for NativeAnimated on iOS

### PR DESCRIPTION
**Pull request by ryangomba**
This diff adds support for native spring animations on iOS. This overlaps some spring work done by @kmagiera on the Android side of things.

**Test plan (required)**

Run UIExplorer NativeAnimated examples before and after - compare the results. Pay special attention to the spring examples.

### CODE REVIEW

**Comment by facebook-github-bot-2** [comment]
By analyzing the blame information on this pull request, we identified @buba447 and @spicyj to be potential reviewers.

**Comment by facebook-github-bot-2** [comment]
Thank you for your pull request.  As you may know, we require contributors to sign our Contributor License Agreement, and we don't seem to have you on file and listed as active anymore.  In order for us to review and merge your code, please email cla@fb.com with your details so we can update your status.

**Comment by facebook-github-bot-1** [comment]
@ryangomba updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@ryangomba updated the pull request.

**Comment by oblador** [comment]
Related PR: https://github.com/facebook/react-native/pull/8860

**Comment by facebook-github-bot-9** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by javache** [comment]
cc @ritzau

**Comment by ryangomba** [comment]
cc @janicduplessis. Do you know if anyone else is actively contributing to iOS native animations? (doesn't look like it, from the commit log). You're my go-to right now :)

**Comment by janicduplessis** [comment]
@javache Do you want someone from FB to have a look at the NativeAnimated PRs before shipping it? If not I can take care of them.

**Comment by javache** [comment]
@janicduplessis @ritzau is ramping up on a lot of the Animated stuff, he's out this week, so I'd prefer if he could take a look at this as well.

**Comment by janicduplessis** [comment]
@ryangomba Could you separate this into 4 different PRs? It will make the review process easier.

File: Libraries/NativeAnimation/Nodes/RCTInterpolationAnimatedNode.m [hunk 1]
```diff
@@ -15,6 +15,8 @@ @implementation RCTInterpolationAnimatedNode
   __weak RCTValueAnimatedNode *_parentNode;
   NSArray<NSNumber *> *_inputRange;
   NSArray<NSNumber *> *_outputRange;
+  BOOL _clampLeft;
```
Comment (janicduplessis) [inline Libraries/NativeAnimation/Nodes/RCTInterpolationAnimatedNode.m]: Can you change this to work with #9366, you can revert the JS changes and assume that you will always get passed `extrapolateLeft` and `extrapolateRight`. Let's also add support for the `identity` value (returns the input) to have full parity with JS.

**Comment by facebook-github-bot-3** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @buba447 as a potential reviewer. Could you take a look please or cc someone with more context?

**Comment by facebook-github-bot-1** [comment]
@ryangomba updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9048/files/987a7715c741cbcefa8fb765e69d81b4c902ab04..d80d7016ef72692431da277b8f81192d3912f55b)

**Comment by ryangomba** [comment]
@janicduplessis Initial separation complete.

**Comment by brentvatne** [comment]
This is pretty important to be able to offload NavigationExperimental animations, @ritzau any chance you can review this soon? cc @janicduplessis

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 2]
```diff
@@ -480,6 +480,7 @@ class SpringAnimation extends Animation {
     this._toValue = config.toValue;
     this._useNativeDriver = config.useNativeDriver !== undefined ? config.useNativeDriver : false;
     this.__isInteraction = config.isInteraction !== undefined ? config.isInteraction : true;
+    this._useNativeDriver = config.useNativeDriver !== undefined ? config.useNativeDriver : false;
```
Comment (foghina) [inline Libraries/Animated/src/AnimatedImplementation.js]: This isn't needed (see line 481 above).

**Comment by foghina** [comment]
Can we delete `RCTAnimationDriverNode.h` & `RCTAnimationDriverNode.m` now?

File: Libraries/NativeAnimation/Drivers/RCTSpringAnimation.m [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+@property (nonatomic, strong) NSNumber *animationId;
+@property (nonatomic, strong) RCTValueAnimatedNode *valueNode;
+@property (nonatomic, assign) BOOL animationHasBegun;
+@property (nonatomic, assign) BOOL animationHasFinished;
+
+@end
+
+@implementation RCTSpringAnimation
+{
+  NSArray<NSNumber *> *_frames;
```
Comment (foghina) [inline Libraries/NativeAnimation/Drivers/RCTSpringAnimation.m]: Is this used?

File: Libraries/NativeAnimation/RCTNativeAnimatedModule.m [hunk 4]
```diff
@@ -166,9 +173,10 @@ - (dispatch_queue_t)methodQueue
     RCTLogError(@"Not a value node.");
     return;
   }
+
   RCTValueAnimatedNode *valueNode = (RCTValueAnimatedNode *)node;
   valueNode.value = value.floatValue;
-  [valueNode setNeedsUpdate];
+  [_updatedValueNodes addObject:valueNode];
```
Comment (foghina) [inline Libraries/NativeAnimation/RCTNativeAnimatedModule.m]: Why this change? Seems like we never added to `_updatedValueNodes` before? Why remove the call to `setNeedsUpdate`?

**Comment by foghina** [comment]
Will try to get an iOS person to look at this next week.

**Comment by ide** [comment]
This needs to be rebased on top of master.

**Comment by foghina** [comment]
@janicduplessis could you have another look at this one as well? We seem to be pretty short staffed on iOS reviewers :(.

@ryangomba this probably needs a rebase.

**Comment by janicduplessis** [comment]
@foghina Sure, I'll try to look at it this week

File: Libraries/NativeAnimation/Drivers/RCTSpringAnimation.m [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    CGFloat dvdt = (aAcceleration + 2 * (bAcceleration + cAcceleration) + dAcceleration) / 6;
+
+    position += dxdt * step;
+    velocity += dvdt * step;
+  }
+
+  _lastPosition = position;
+  _lastVelocity = velocity;
+
+  if (_animationHasFinished) {
```
Comment (janicduplessis) [inline Libraries/NativeAnimation/Drivers/RCTSpringAnimation.m]: Is this check useful? As far as I can tell we don't update this value and we already check it at the start of the method, save for the one a few lines later and the one at the end.

File: Libraries/NativeAnimation/Drivers/RCTSpringAnimation.m [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    if (_tension != 0) {
+      // Ensure that we end up with a round value
+      if (_animationHasFinished) {
+        return;
+      }
+      [self onUpdate:_toValue];
+    }
+
+    [self stopAnimation];
+    return;
```
Comment (janicduplessis) [inline Libraries/NativeAnimation/Drivers/RCTSpringAnimation.m]: Nit: Can you get rid of this return? Looks odd to me.

File: Libraries/NativeAnimation/Drivers/RCTSpringAnimation.m [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+  CGFloat _tension;
+  CGFloat _friction;
+  CGFloat _initialVelocity;
+  NSTimeInterval _animationStartTime;
+  NSTimeInterval _animationCurrentTime;
+  RCTResponseSenderBlock _callback;
+
+  CGFloat _lastPosition;
+  CGFloat _lastVelocity;
+  CGFloat _outputValue;
```
Comment (janicduplessis) [inline Libraries/NativeAnimation/Drivers/RCTSpringAnimation.m]: I think we can get rid of this, we're only writing to it and never using it's value.

File: Libraries/NativeAnimation/RCTNativeAnimatedModule.m [hunk 8]
```diff
@@ -24,9 +26,9 @@
 @implementation RCTNativeAnimatedModule
 {
   NSMutableDictionary<NSNumber *, RCTAnimatedNode *> *_animationNodes;
-  NSMutableDictionary<NSNumber *, RCTAnimationDriverNode *> *_animationDrivers;
-  NSMutableSet<RCTAnimationDriverNode *> *_activeAnimations;
-  NSMutableSet<RCTAnimationDriverNode *> *_finishedAnimations;
+  NSMutableDictionary<NSNumber *, NSObject<RCTAnimationDriver> *> *_animationDrivers;
```
Comment (janicduplessis) [inline Libraries/NativeAnimation/RCTNativeAnimatedModule.m]: Nit: Can you use `id<...>` instead of `NSObject<...>` in this file.

File: Libraries/NativeAnimation/RCTNativeAnimatedModule.m (hunk 4 above)
Comment (janicduplessis) [inline Libraries/NativeAnimation/RCTNativeAnimatedModule.m]: Yea, not sure about this change either, looks like `_updatedValueNodes` is pretty useless at the moment. We are only calling cleanupAnimationUpdates on it during the animation loop.

I did notice that `setAnimatedNodeValue` was broken anyway and didn't actually update the value if there is no other animation running. Might want to just revert this and fix it separatly.

**Comment by janicduplessis** [comment]
@ryangomba This is one of the last big missing part for native animations, let me know if you plan to have time to work on this soon or if you'd like you can give maintainer access to your branch and I can do the final adjustments before landing this.

**Comment by ryangomba** [comment]
I'll follow up on this today

**Comment by ryangomba** [comment]
@janicduplessis I've addressed the issues caught above; this should be good to go!

**Comment by janicduplessis** [comment]
@ryangomba Looks like the build is failing for TVOS, it was added pretty recently and I don't know much about it. https://travis-ci.org/facebook/react-native/jobs/169298841#L6268. Maybe @dlowder-salesforce can help here.

**Comment by ryangomba** [comment]
@janicduplessis should pass this time

**Comment by janicduplessis** [comment]
@ryangomba One test is still failing but it looks unrelated probably just flacky. Can you rebase on master there is a conflict now because I merged your clamp PR.

**Comment by dlowder-salesforce** [comment]
@janicduplessis @ryangomba the Apple TV issue is happening because RCTFrameAnimation.m and RCTSpringAnimation.m are included in the iOS target, but not in the tvOS target.  So UIExplorer fails to link due to missing object files for those classes.

**Comment by ryangomba** [comment]
@dlowder-salesforce right, it's fixed now

**Comment by janicduplessis** [comment]
Great! Thanks for your work on this @ryangomba, going to wait for CI then shipit.

**Comment by janicduplessis** [comment]
@facebook-github-bot shipit
